lib/Common/DataStructures/ClusterList.h

```diff
@@ -0,0 +1,402 @@
+ * the array in cases where we have large functions (which is more common than code
+ * input may indicate due to inlining), assuming that the data structure would need
+ * to be allocated on a per-block basis otherwise.
+ */
+
+template<typename indexType, class TAllocator, unsigned int numPerSegment = 16>
+class SegmentClusterList
+{
+private:
+    // The maximum index enabled; basically an array bounds for the set.
+    indexType maxIndex;
+    // The backing store for this ClusterList type is an array of pointers to subsegment
+    // structures, each of which stores a fixed region. Elements outside this region are
+    // assumed to have an identity mapping before first use.
+    indexType **backingStore;
+#if DBG
+    // During the merge process, we may end up with some situations where set membership
+    // is chained; that is, that node 7 is in set 5, but 5 is in set 4. Following all of
+    // the merge operations we have a step to reconcile all of this information and then
+    // update each element of the list to point to the least member of its set. This can
+    // be checked in debug builds to see if we've done this step or not.
+    bool consolidated = false;
+#endif
+    TAllocator* alloc;
+    inline void EnsureBaseSize(indexType inputMax)
+    {
+        // We want to ensure that we can store to inputMax, which means we need to grow to inputMax+1
+        indexType targetSize = inputMax + 1;
+        // round up to the next numPerSegment multiple
+        indexType max = (targetSize % numPerSegment == 0) ? targetSize : (targetSize + numPerSegment - (targetSize % numPerSegment));
+        // get segment counts for allocation
+        indexType newNumSegments = max / numPerSegment;
+        indexType curNumSegments = maxIndex / numPerSegment;
+        if (newNumSegments > curNumSegments)
+        {
+            // Note that if you want to use this with something other than an ArenaAllocator, then
+            // you'll have to add clean-up mechanics
+            indexType **tempList = AllocatorNewArray(TAllocator, alloc, indexType*, newNumSegments);
+            for (indexType i = 0; i < curNumSegments; i++)
+            {
+                tempList[i] = backingStore[i];
+            }
+            for (indexType i = curNumSegments; i < newNumSegments; i++)
+            {
+                tempList[i] = nullptr;
+            }
+            // Set the max index to the number of supported indicies
+            maxIndex = max;
+            // replace the backingStore with the new backing Store
+            indexType **oldBackingStore = backingStore;
+            backingStore = tempList;
+            AllocatorDeleteArray(TAllocator, alloc, curNumSegments, oldBackingStore);
+        }
+    }
+    inline void CreateBacking(indexType index)
+    {
+        Assert(index >= maxIndex || backingStore[index / numPerSegment] == nullptr);
+        // grow the size of the pointer array if needed
+        EnsureBaseSize(index + 1);
```

Comment:
Looks like EnsureBaseSize already adds one internally; do we need to add another here?

Reply:
No, we don't. Fixing in next commit.